File: ui-tests/PageObjects/Panes/EventLog/EventLogPane.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using Microsoft.Playwright;
using UiTests.PageObjects;
using UiTests.Utils;

namespace UiTests.PageObjects.Panes.EventLog;

/// <summary>
/// Event log pane containing multiple event rows.
/// </summary>
public class EventLogPane : TabObject
{
    private List<EventRow> _events = new();

    public EventLogPane(IPage page, ILocator root, string tabButtonText)
        : base(page, root, tabButtonText)
    {
    }

    public ILocator FooterContainer() => Root.Locator(".data-tables-footer");

    public ILocator RowsInfoContainer() => FooterContainer().Locator(".data-tables-footer-info");

    public async Task<int> RowsAsync()
    {
        var klass = await FooterContainer().GetAttributeAsync("class");
        var m = Regex.Match(klass ?? string.Empty, @"(\d*)to");
        return m.Success ? int.Parse(m.Groups[1].Value) : 0;
        }

    public async Task<int> ToRowAsync()
    {
        var text = await RowsInfoContainer().TextContentAsync();
        var m = Regex.Match(text ?? string.Empty, @"(\d*)\sof");
        return m.Success ? int.Parse(m.Groups[1].Value) : 0;
    }

    public async Task<int> OfRowsAsync()
    {
        var text = await RowsInfoContainer().TextContentAsync();
        var m = Regex.Match(text ?? string.Empty, @"of\s(\d*)");
        return m.Success ? int.Parse(m.Groups[1].Value) : 0;
    }

    private async Task<IReadOnlyList<ILocator>> EventElementRootsAsync()
        => await Root.Locator(".eventLog-dense-table tbody tr").AllAsync();

    public async Task<IReadOnlyList<EventRow>> EventElementsAsync(bool forceReload = false)
    {
        if (forceReload || _events.Count == 0)
        {
            var roots = await EventElementRootsAsync();
            _events = roots.Select(r => new EventRow(r, EventElementType.EventLog)).ToList();
        }
        return _events;
    }

    public Task<int> RowCountAsync(bool forceReload = false)
        => Root.Locator(".eventLog-dense-table tbody tr").CountAsync();

    public async Task<EventRow> RowByIndexAsync(int index, bool forceReload = false)
    {
        DebugLogger.Log($"EventLogPane: locating row {index} (forceReload={forceReload})");
        var rows = await EventElementsAsync(forceReload);
        foreach (var row in rows)
        {
            if (await row.IndexAsync() == index)
            {
                DebugLogger.Log($"EventLogPane: found row {index}");
                return row;
            }
        }

        DebugLogger.Log($"EventLogPane: row {index} not found");
        throw new InvalidOperationException($"Event log row with index {index} was not found.");
    }

    private ILocator FilterButton()
        => Root.GetByText("Filter", new() { Exact = true }).First;

    private ILocator DropdownRoot()
        => Page.Locator("#dropdown-container-id");

    public async Task ActivateTraceEventsFilterAsync()
    {
        await FilterButton().ClickAsync();
        var traceButton = DropdownRoot().GetByText("Trace events", new() { Exact = true });
        await traceButton.WaitForAsync(new() { State = WaitForSelectorState.Visible });
        await traceButton.ClickAsync();
        await Page.Keyboard.PressAsync("Escape");
        await Page.WaitForTimeoutAsync(100);
    }

    public async Task ActivateRecordedEventsFilterAsync()
    {
        await FilterButton().ClickAsync();
        var recordedButton = DropdownRoot().GetByText("Recorded events", new() { Exact = true });
        await recordedButton.WaitForAsync(new() { State = WaitForSelectorState.Visible });
        await recordedButton.ClickAsync();
        await Page.Keyboard.PressAsync("Escape");
        await Page.WaitForTimeoutAsync(100);
    }
}

File: tsc-ui-tests/page-objects/panes/event-log/event-log-pane.ts
import type { Locator, Page } from "@playwright/test";
import { EventRow, EventElementType } from "./event-row";
import { debugLogger } from "../../../lib/debug-logger";

const POST_FILTER_DELAY_MS = 100;

/**
 * Event log pane containing multiple event rows.
 *
 * Port of ui-tests/PageObjects/Panes/EventLog/EventLogPane.cs
 */
export class EventLogPane {
  readonly page: Page;
  readonly root: Locator;
  readonly tabButtonText: string;
  private events: EventRow[] = [];

  constructor(page: Page, root: Locator, tabButtonText: string) {
    this.page = page;
    this.root = root;
    this.tabButtonText = tabButtonText;
  }

  tabButton(): Locator {
    return this.page.locator(".lm_title", { hasText: this.tabButtonText }).first();
  }

  async isVisible(): Promise<boolean> {
    const style = await this.root.locator("..").getAttribute("style");
    return !(style?.includes("none"));
  }

  footerContainer(): Locator {
    return this.root.locator(".data-tables-footer");
  }

  rowsInfoContainer(): Locator {
    return this.footerContainer().locator(".data-tables-footer-info");
  }

  async rows(): Promise<number> {
    const klass = (await this.footerContainer().getAttribute("class")) ?? "";
    const m = klass.match(/(\d*)to/);
    return m ? parseInt(m[1], 10) || 0 : 0;
  }

  async toRow(): Promise<number> {
    const text = (await this.rowsInfoContainer().textContent()) ?? "";
    const m = text.match(/(\d*)\sof/);
    return m ? parseInt(m[1], 10) || 0 : 0;
  }

  async ofRows(): Promise<number> {
    const text = (await this.rowsInfoContainer().textContent()) ?? "";
    const m = text.match(/of\s(\d*)/);
    return m ? parseInt(m[1], 10) || 0 : 0;
  }

  private eventElementRoots(): Locator {
    return this.root.locator(".eventLog-dense-table tbody tr");
  }

  async eventElements(forceReload = false): Promise<EventRow[]> {
    if (forceReload || this.events.length === 0) {
      const roots = await this.eventElementRoots().all();
      this.events = roots.map((r) => new EventRow(r, EventElementType.EventLog));
    }
    return this.events;
  }

  async rowCount(): Promise<number> {
    return this.eventElementRoots().count();
  }

  async rowByIndex(index: number, forceReload = false): Promise<EventRow> {
    debugLogger.log(`EventLogPane: locating row ${index} (forceReload=${forceReload})`);
    const rows = await this.eventElements(forceReload);
    for (const row of rows) {
      if ((await row.index()) === index) {
        debugLogger.log(`EventLogPane: found row ${index}`);
        return row;
      }
    }
    debugLogger.log(`EventLogPane: row ${index} not found`);
    throw new Error(`Event log row with index ${index} was not found.`);
  }

  private filterButton(): Locator {
    return this.root.getByText("Filter", { exact: true }).first();
  }

  private dropdownRoot(): Locator {
    return this.page.locator("#dropdown-container-id");
  }

  async activateTraceEventsFilter(): Promise<void> {
    await this.filterButton().click();
    const traceButton = this.dropdownRoot().getByText("Trace events", { exact: true });
    await traceButton.waitFor({ state: "visible" });
    await traceButton.click();
    await this.page.keyboard.press("Escape");
    await this.page.waitForTimeout(POST_FILTER_DELAY_MS);
  }

  async activateRecordedEventsFilter(): Promise<void> {
    await this.filterButton().click();
    const recordedButton = this.dropdownRoot().getByText("Recorded events", { exact: true });
    await recordedButton.waitFor({ state: "visible" });
    await recordedButton.click();
    await this.page.keyboard.press("Escape");
    await this.page.waitForTimeout(POST_FILTER_DELAY_MS);
  }
}
